packages/transaction-controller/src/helpers/TransactionPoller.ts

```diff
@@ -0,0 +1,168 @@
+
+  #pendingTransactions?: TransactionMeta[];
+
+  #running = false;
+
+  #timeout?: NodeJS.Timeout;
+
+  constructor(blockTracker: BlockTracker) {
+    this.#blockTracker = blockTracker;
+  }
+
+  /**
+   * Start the poller with a listener that will be called on every interval.
+   * @param listener - The listener to call on every interval.
+   */
+  start(listener: (latestBlockNumber: string) => Promise<void>) {
+    if (this.#running) {
+      return;
+    }
+
+    this.#listener = listener;
+    this.#running = true;
+
+    this.#queue();
+
+    log('Started');
+  }
+
+  /**
+   * Stop the poller.
+   * Remove all timeouts and block tracker listeners.
+   */
+  stop() {
+    if (!this.#running) {
+      return;
+    }
+
+    this.#running = false;
+    this.#listener = undefined;
+    this.#acceleratedCount = 0;
+    this.#pendingTransactions = undefined;
+
+    this.#stopTimeout();
+    this.#stopBlockTracker();
+
+    log('Stopped');
+  }
+
+  /**
+   * Notify the poller of the pending transactions being monitored.
+   * This will reset to the accelerated polling and reset the count
+   * when new transactions are added or removed.
+   * @param pendingTransactions - The pending transactions to poll.
+   */
+  setPendingTransactions(pendingTransactions: TransactionMeta[]) {
+    const currentPendingTransactionIds = (this.#pendingTransactions ?? []).map(
+      (tx) => tx.id,
+    );
+
+    this.#pendingTransactions = pendingTransactions;
+
+    const newPendingTransactionIds = pendingTransactions.map((tx) => tx.id);
+
+    const hasUpdatedIds = !isEqual(
+      currentPendingTransactionIds,
+      newPendingTransactionIds,
+    );
+
+    if (!this.#running || !hasUpdatedIds) {
+      return;
+    }
+
+    log('Detected new pending transactions', newPendingTransactionIds);
+
+    this.#acceleratedCount = 0;
+
+    if (this.#blockTrackerListener) {
+      this.#stopBlockTracker();
+      this.#queue();
+    }
+  }
+
+  #queue() {
+    if (!this.#running) {
+      return;
+    }
+
+    if (this.#acceleratedCount >= ACCELERATED_COUNT_MAX) {
+      // eslint-disable-next-line @typescript-eslint/no-misused-promises
+      this.#blockTrackerListener = (latestBlockNumber) =>
+        this.#interval(false, latestBlockNumber);
+
+      this.#blockTracker.on('latest', this.#blockTrackerListener);
+
+      log('Added block tracker listener');
+
+      return;
+    }
+
+    this.#stopTimeout();
+
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    this.#timeout = setTimeout(async () => {
+      await this.#interval(true);
+      this.#queue();
+    }, ACCELERATED_INTERVAL);
+  }
+
+  async #interval(isAccelerated: boolean, latestBlockNumber?: string) {
+    if (isAccelerated) {
+      log('Accelerated interval', this.#acceleratedCount + 1);
+    } else {
+      log('Block tracker interval', latestBlockNumber);
+    }
+
+    const latestBlockNumberFinal =
+      latestBlockNumber ?? (await this.#blockTracker.getLatestBlock());
```

Comment:
The latest block we're using is just the latest one that the block tracker knows about, so we would expect this to be delayed, right?

We have block cache middleware that caches the response to `eth_getTransactionReceipt` by block (see https://github.com/MetaMask/eth-json-rpc-middleware/blob/bf4fe7395728bbdc839d0cb68e920e84f9e5ba9b/src/utils/cache.ts#L163, and note that all cache types are resolved as block caches [here](https://github.com/MetaMask/eth-json-rpc-middleware/blob/bf4fe7395728bbdc839d0cb68e920e84f9e5ba9b/src/block-cache.ts#L152)). Meaning, this extra polling might not result in any extra requests, just unnecessary no-op cache hits.

Reply:
Correct, this is just to satisfy the existing reliance on the block number in the `PendingTransactionTracker`.

It's okay for this to return a cached value.

Reply:
Following investigation, I can confirm this is not hitting the block cache since `eth_getTransactionReceipt` returns a null result which are not cached.